### PR TITLE
Adding loop-over-history functionality to replay-tx script

### PR DESCRIPTION
**Pull request by gigamesh**
Attempting to finish the work described here: https://github.com/ethereum-optimism/optimism/issues/705

- update watcher to return failed txs using new event referenced here: https://github.com/ethereum-optimism/optimism/issues/602
- get all failed txs from provided `blockNumber` & replay them

### CODE REVIEW

File: scripts/replay-tx.js [hunk 1]
```diff
@@ -7,25 +12,43 @@ const { getContractInterface } = require("@eth-optimism/contracts");
...
-  const receipt = await mainnetProvider.getTransactionReceipt(argOptions.hash);
+  if (!argOptions.hash && !argOptions.blockNumber) {
+    console.error("Must provide a hash or block number.");
+    return;
+  }
+
+  if (argOptions.hash) {
+    replayMessage(argOptions.hash);
+  } else {
+    // TODO: loop over blocks starting at argOptions.blockNumber, get all failed transactions and replay them
```
Comment (gigamesh) [inline scripts/replay-tx.js]: 🤷‍♂️

Comment (tynes) [reply scripts/replay-tx.js]: Are you planning on adding this in this PR?

Comment (gigamesh) [reply scripts/replay-tx.js]: yep! Liam & Kevin unblocked me. thanks

File: scripts/replay-tx.js [hunk 2]
```diff
@@ -1,31 +1,73 @@
...
+  if (argOptions.hash) {
+    replayMessage(argOptions.hash);
+  } else {
+    // Replay all transactions from provided blockNumber
+    try {
+      const logs = await provider.getLogs({
+        address: messengerAddress,
+        topics: [utils.id(`RelayedMessage(bytes32)`)],
+        fromBlock: Number(argOptions.blockNumber,
+      });
```
Comment (K-Ho) [inline scripts/replay-tx.js]: If this is connected to the L1 provider and L1 messenger, you should be looking for `SentMessage` events - replay all SentMessage (deposits) txs

File: scripts/replay-tx.js [hunk 3]
```diff
@@ -1,31 +1,73 @@
...
-const { JsonRpcProvider, getDefaultProvider } = require("@ethersproject/providers");
+const { JsonRpcProvider, InfuraProvider } = require("@ethersproject/providers");
 const program = new Command();
 
-program.requiredOption("-h, --hash <number>", "specify transaction hash");
 program.requiredOption("-n, --network <string>", "specify network");
-program.option("-k, --key <number>", "specify private key");
+program.requiredOption("-k, --key <number>", "specify private key");
+program.option("-h, --hash <number>", "specify transaction hash");
+program.option("-b, --blockNumber <number>", "specify a block number");
```
Comment (karlfloersch) [inline scripts/replay-tx.js]: Can we also add a `toBlock` in addition to the `fromBlock` ?

**Comment by snario** [comment]
cc @karlfloersch for final review

File: scripts/replay-tx.js [hunk 4]
```diff
@@ -1,31 +1,97 @@
...
+
+if (!argOptions.hash && !argOptions.fromBlock) {
+  console.error("Must provide a hash or block number to start from.");
+  return;
+}
+const provider = new InfuraProvider(argOptions.network, process.env.INFURA_KEY);
 
-const L1_MESSENGER_PROXY = "0xD1EC7d40CCd01EB7A305b94cBa8AB6D17f6a9eFE";
+// Addresses from May 11 regenesis
+const MAINNET_L1_MESSENGER_PROXY = "0x902e5fF5A99C4eC1C21bbab089fdabE32EF0A5DF";
```
Comment (smartcontracts) [inline scripts/replay-tx.js]: Need to make sure these are up to date.

File: scripts/replay-tx.js [hunk 5]
```diff
@@ -1,31 +1,97 @@
...
+const proxyL1Messenger = new Contract(messengerAddress, getContractInterface("OVM_L1CrossDomainMessenger"), wallet);
 
 const main = async () => {
-  const receipt = await mainnetProvider.getTransactionReceipt(argOptions.hash);
+  if (argOptions.clearPendingTx) {
+    latestTxCount = await wallet.getTransactionCount("latest");
+    pendingTxCount = await wallet.getTransactionCount("pending");
+    if (latestTxCount < pendingTxCount) {
+      console.log("Detected a pending transaction! Clearing it...");
+      await wallet.sendTransaction({
```
Comment (smartcontracts) [inline scripts/replay-tx.js]: You should `.wait()` the result of this

File: scripts/replay-tx.js [hunk 6]
```diff
@@ -1,31 +1,97 @@
...
+        to: await wallet.getAddress(),
+        value: 0,
+        nonce: latestTxCount,
+      });
+    }
+    if (latestTxCount + 1 < pendingTxCount) {
```
Comment (smartcontracts) [inline scripts/replay-tx.js]: Do you know if there's any way to clear out /all/ pending transactions at the same time?

File: scripts/replay-tx.js [hunk 7]
```diff
@@ -1,31 +1,97 @@
...
+      );
+      process.exit(1);
+    }
+  }
+  if (argOptions.hash) {
+    replayMessage(argOptions.hash);
+  } else {
+    // Replay all transactions from provided toBlock number
+    try {
+      const logParams = {
```
Comment (smartcontracts) [inline scripts/replay-tx.js]: It's cleaner to use `messenger.queryFilter`

Comment (smartcontracts) [reply scripts/replay-tx.js]: In combination with `messenger.filters.SentMessage()`

File: scripts/replay-tx.js [hunk 8]
```diff
@@ -1,31 +1,97 @@
...
+        address: messengerAddress,
+        topics: [utils.id(`SentMessage(bytes)`)],
+        fromBlock: Number(argOptions.fromBlock),
+      };
+      if (argOptions.toBlock) {
+        logParams.toBlock = Number(argOptions.toBlock);
+      }
+      const logs = await provider.getLogs(logParams);
+      const txHashes = logs.map((log) => log.transactionHash);
+      for (const txHash of txHashes) {
```
Comment (smartcontracts) [inline scripts/replay-tx.js]: I think more legible to loop over the logs and then pass on log.transactionHash

**Review by smartcontracts — CHANGES_REQUESTED** [review]
Requesting a few small updates

**Review by smartcontracts — APPROVED** [review]

**Pull request merged**